### ISSUE TITLE
CMake is unable to `find_package(Ruby)` because of unusual `with-soname` in ruby package

### ISSUE THREAD

**Issue by jiridanek** [NixOS/nixpkgs#85001]
CMake has certain expectations when running `FindRuby.cmake`. The ruby nixpkg does not meet those. The problem is that the `.so` file is named `libruby_v2_6_5.so` instead of `libruby-2.6.5.so`, which is one of the values in `_RUBY_POSSIBLE_LIB_NAMES` as shown below (one of values in `_RUBY_POSSIBLE_LIB_NAMES` CMake variable).

**To Reproduce**

```cmake
set(_RUBY_DEBUG_OUTPUT TRUE)
find_package(Ruby)
```

This prints

```
-- --------FindRuby.cmake debug------------
-- _RUBY_POSSIBLE_EXECUTABLE_NAMES: ruby;ruby2.4;ruby24;ruby2.3;ruby23;ruby2.2;ruby22;ruby2.1;ruby21;ruby2.0;ruby20;ruby1.9;ruby19;ruby1.8;ruby18
-- _RUBY_POSSIBLE_LIB_NAMES: ruby;ruby-static;ruby2.6;ruby26;ruby-2.6;ruby-2.6.5
-- RUBY_ARCH_DIR: /nix/store/idjl5n00a8wr16zhwn3mrk1f3qr4bxkx-ruby-2.6.5/lib/ruby/2.6.0/x86_64-linux
-- RUBY_HDR_DIR: /nix/store/idjl5n00a8wr16zhwn3mrk1f3qr4bxkx-ruby-2.6.5/include/ruby-2.6.0
-- RUBY_POSSIBLE_LIB_DIR: /nix/store/idjl5n00a8wr16zhwn3mrk1f3qr4bxkx-ruby-2.6.5/lib
-- Found RUBY_VERSION: "2.6.5" , short: "2.6", nodot: "26"
-- _RUBY_REQUIRED_VARS: RUBY_EXECUTABLE;RUBY_INCLUDE_DIR;RUBY_LIBRARY;RUBY_CONFIG_INCLUDE_DIR
-- RUBY_EXECUTABLE: /nix/store/idjl5n00a8wr16zhwn3mrk1f3qr4bxkx-ruby-2.6.5/bin/ruby
-- RUBY_LIBRARY: RUBY_LIBRARY-NOTFOUND
-- RUBY_INCLUDE_DIR: /nix/store/idjl5n00a8wr16zhwn3mrk1f3qr4bxkx-ruby-2.6.5/include/ruby-2.6.0
-- RUBY_CONFIG_INCLUDE_DIR: /nix/store/idjl5n00a8wr16zhwn3mrk1f3qr4bxkx-ruby-2.6.5/include/ruby-2.6.0/x86_64-linux
-- --------------------
-- Could NOT find Ruby (missing: RUBY_LIBRARY) (found version "2.6.5")
```

```bash
$ ls /nix/store/idjl5n00a8wr16zhwn3mrk1f3qr4bxkx-ruby-2.6.5/liblibruby_v2_6_5.so
libruby_v2_6_5.so.2.6  libruby_v2_6_5.so.2.6.5  pkgconfig  ruby
```

For example, `FindPython` works fine. The libraries there are

```
$ ls /nix/store/n8nviwmllwqv0fjsar8v8k8gjap1vhcw-python3-3.7.6/lib/
libpython3.7m.so  libpython3.7m.so.1.0  libpython3.so  pkgconfig  python3.7
```

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.22, NixOS, 20.03beta321.be346a1f4bd (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.2`
 - channels(root): `"nixos-20.03beta321.be346a1f4bd, nixos-hardware"`
 - channels(jdanek): `"nixos-unstable-20.09pre218613.ae6bdcc5358"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by jiridanek** [comment]
I am looking at the following, with suspicion,

https://github.com/NixOS/nixpkgs/blob/84cf00f98031e93f389f1eb93c4a7374a33cc0a9/pkgs/development/interpreters/ruby/default.nix#L118-L120

CC maintainers @vrthra @manveru

**Comment by marsam** [comment]
cc: @Izorkin https://github.com/NixOS/nixpkgs/pull/73491/commits/7d0fcb8acc282577aa1bf5fb9bb91e9359f4d6b3

**Comment by Izorkin** [comment]
With this variant worked?
```
diff --git a/pkgs/development/interpreters/ruby/default.nix b/pkgs/development/interpreters/ruby/default.nix
index 1756c13b53d..faad826e46a 100644
--- a/pkgs/development/interpreters/ruby/default.nix
+++ b/pkgs/development/interpreters/ruby/default.nix
@@ -116,7 +116,7 @@ let
         # https://github.com/ruby/ruby/commit/97a5af62a318fcd93a4e5e4428d576c0280ddbae
         buildFlags = lib.optionals atLeast27 [ "REVISION_LATEST=0" ];

-        configureFlags = ["--enable-shared" "--enable-pthread" "--with-soname=ruby_${tag}"]
+        configureFlags = ["--enable-shared" "--enable-pthread" "--with-soname=ruby-${tag}"]
           ++ op useRailsExpress "--with-baseruby=${baseruby}/bin/ruby"
           ++ op (!docSupport) "--disable-install-doc"
           ++ ops stdenv.isDarwin [
diff --git a/pkgs/servers/sql/mariadb/default.nix b/pkgs/servers/sql/mariadb/default.nix
index ea11faf1efc..d3cc08eaadd 100644
```

**Comment by jiridanek** [comment]
@Izorkin I think not. Assuming that `--with-soname` really sets the filename of .so files, and not just the soname filed in the .so ELF object (something I did not check), then there still would be the leading `v` and underscores where periods are expected by CMake. That is, `libruby-v2_6_5.so` instead of `libruby-2.6.5.so`

**Comment by Izorkin** [comment]
```
-- Running cmake version 3.16.5
-- --------FindRuby.cmake debug------------
-- _RUBY_POSSIBLE_EXECUTABLE_NAMES: ruby;ruby2.4;ruby24;ruby2.3;ruby23;ruby2.2;ruby22;ruby2.1;ruby21;ruby2.0;ruby20;ruby1.9;ruby19;ruby1.8;ruby18
-- _RUBY_POSSIBLE_LIB_NAMES: ruby;ruby-static;ruby2.6;ruby26;ruby-2.6;ruby-2.6.6
-- RUBY_ARCH_DIR: /nix/store/9zbif4pzi5qkdhh8sqn87w11jvw95r42-ruby-2.6.6/lib/ruby/2.6.0/x86_64-linux
-- RUBY_HDR_DIR: /nix/store/9zbif4pzi5qkdhh8sqn87w11jvw95r42-ruby-2.6.6/include/ruby-2.6.0
-- RUBY_POSSIBLE_LIB_DIR: /nix/store/9zbif4pzi5qkdhh8sqn87w11jvw95r42-ruby-2.6.6/lib
-- Found RUBY_VERSION: "2.6.6" , short: "2.6", nodot: "26"
-- _RUBY_REQUIRED_VARS: RUBY_EXECUTABLE;RUBY_INCLUDE_DIR;RUBY_LIBRARY;RUBY_CONFIG_INCLUDE_DIR
-- RUBY_EXECUTABLE: /nix/store/9zbif4pzi5qkdhh8sqn87w11jvw95r42-ruby-2.6.6/bin/ruby
-- RUBY_LIBRARY: /nix/store/9zbif4pzi5qkdhh8sqn87w11jvw95r42-ruby-2.6.6/bin/ruby
-- RUBY_INCLUDE_DIR: /nix/store/9zbif4pzi5qkdhh8sqn87w11jvw95r42-ruby-2.6.6/include/ruby-2.6.0
-- RUBY_CONFIG_INCLUDE_DIR: /nix/store/9zbif4pzi5qkdhh8sqn87w11jvw95r42-ruby-2.6.6/include/ruby-2.6.0/x86_64-linux
-- --------------------
```

**Comment by jiridanek** [comment]
```RUBY_LIBRARY: /nix/store/9zbif4pzi5qkdhh8sqn87w11jvw95r42-ruby-2.6.6/bin/ruby```

This looks wrong. It should find the .so file there, not the interpreter binary.

**Comment by Izorkin** [comment]
with 
```
diff --git a/pkgs/development/interpreters/ruby/default.nix b/pkgs/development/interpreters/ruby/default.nix
index 1756c13b53d..dbfea17437d 100644
--- a/pkgs/development/interpreters/ruby/default.nix
+++ b/pkgs/development/interpreters/ruby/default.nix
@@ -116,7 +116,7 @@ let
         # https://github.com/ruby/ruby/commit/97a5af62a318fcd93a4e5e4428d576c0280ddbae
         buildFlags = lib.optionals atLeast27 [ "REVISION_LATEST=0" ];

-        configureFlags = ["--enable-shared" "--enable-pthread" "--with-soname=ruby_${tag}"]
+        configureFlags = ["--enable-shared" "--enable-pthread" "--with-soname=ruby-${version}"]
           ++ op useRailsExpress "--with-baseruby=${baseruby}/bin/ruby"
           ++ op (!docSupport) "--disable-install-doc"
           ++ ops stdenv.isDarwin [
```

```
-- Running cmake version 3.16.5
-- --------FindRuby.cmake debug------------
-- _RUBY_POSSIBLE_EXECUTABLE_NAMES: ruby;ruby2.4;ruby24;ruby2.3;ruby23;ruby2.2;ruby22;ruby2.1;ruby21;ruby2.0;ruby20;ruby1.9;ruby19;ruby1.8;ruby18
-- _RUBY_POSSIBLE_LIB_NAMES: ruby;ruby-static;ruby2.6;ruby26;ruby-2.6;ruby-2.6.6
-- RUBY_ARCH_DIR: /nix/store/5nvmryah49jrzqcsx07v5dbchhzybznk-ruby-2.6.6/lib/ruby/2.6.0/x86_64-linux
-- RUBY_HDR_DIR: /nix/store/5nvmryah49jrzqcsx07v5dbchhzybznk-ruby-2.6.6/include/ruby-2.6.0
-- RUBY_POSSIBLE_LIB_DIR: /nix/store/5nvmryah49jrzqcsx07v5dbchhzybznk-ruby-2.6.6/lib
-- Found RUBY_VERSION: "2.6.6" , short: "2.6", nodot: "26"
-- _RUBY_REQUIRED_VARS: RUBY_EXECUTABLE;RUBY_INCLUDE_DIR;RUBY_LIBRARY;RUBY_CONFIG_INCLUDE_DIR
-- RUBY_EXECUTABLE: /nix/store/5nvmryah49jrzqcsx07v5dbchhzybznk-ruby-2.6.6/bin/ruby
-- RUBY_LIBRARY: /nix/store/5nvmryah49jrzqcsx07v5dbchhzybznk-ruby-2.6.6/lib/libruby-2.6.6.so
-- RUBY_INCLUDE_DIR: /nix/store/5nvmryah49jrzqcsx07v5dbchhzybznk-ruby-2.6.6/include/ruby-2.6.0
-- RUBY_CONFIG_INCLUDE_DIR: /nix/store/5nvmryah49jrzqcsx07v5dbchhzybznk-ruby-2.6.6/include/ruby-2.6.0/x86_64-linux
-- --------------------

```

**Comment by jiridanek** [comment]
Thank you for looking into this. Yes, now it looks correct to me.

**Comment by jiridanek** [comment]
I had a look how Ubuntu does it. They have `/usr/lib/x86_64-linux-gnu/libruby-2.7.so`, that is, they drop the minor number from filename. Doing that would only complicate things in nixpkgs and it would not help anything, so no need to follow them exactly. https://packages.ubuntu.com/focal/amd64/ruby2.7-dev/filelist

**Comment by Izorkin** [comment]
Fixed in #87239
